hypershift-operator/controllers/nodepool/inplace.go

```diff
@@ -66,27 +68,29 @@ func (r *NodePoolReconciler) reconcileInPlaceUpgrade(ctx context.Context, hc *hy
 		return nil
 	}
 
-	hostedClusterClient, err := newHostedClusterClient(ctx, r.Client, hc)
-	if err != nil {
-		return fmt.Errorf("failed to create remote client: %w", err)
-	}
-
 	// Watch hosted cluster Nodes. We track the created caches, so we don't add a watcher on every reconciliation.
-	// TODO (alberto): cache by HC instead so we reduce the cache size.
+	// This cache might be accessed in parallel by multiple reconciliations on the same cache key if
+	// a hostedcluster has multiple nodepoool, thus we do not use a RWLock.
 	r.hostedClusterCachesTracker.Lock()
 	defer r.hostedClusterCachesTracker.Unlock()
-	if _, ok := r.hostedClusterCachesTracker.caches[client.ObjectKeyFromObject(nodePool)]; !ok {
-		hostedClusterCache, err := newHostedClusterCache(ctx, r.Client, hc)
+	cacheEntry, ok := r.hostedClusterCachesTracker.caches[client.ObjectKeyFromObject(hc)]
+	if !ok {
+		hostedClusterCache, hostedClusterClient, err := newHostedClusterCacheAndClient(ctx, r.Client, hc)
 		if err != nil {
 			return fmt.Errorf("failed to create hosted cluster cache: %w", err)
 		}
 
 		cacheCtx, cacheCtxCancel := context.WithCancel(ctx)
 		// TODO: index by HC here instead?
```

Comment:
this TODO (seems dup of https://github.com/openshift/hypershift/pull/1358/files#diff-186fca173b913fab90bc6a26071e1255532f14ce6156a796e04de919d8e97c13L75) can go away now as well.